table/tables/tables.go

```diff
@@ -582,6 +633,76 @@ func (t *TableCommon) rebuildIndices(ctx sessionctx.Context, txn kv.Transaction,
 	return nil
 }
 
+// checksumPrecheck checks if we need/can write row with checksums. The first returned int indicates the number of
+// checksums we need, the second returned column (when it's not nil) is the original column of modify-column ddl.
+func (t *TableCommon) checksumPrecheck(sctx sessionctx.Context) (int, *table.Column, error) {
+	var (
+		origCol    *table.Column
+		nonPubCols []*model.ColumnInfo
+	)
+	if !sctx.GetSessionVars().IsRowLevelChecksumEnabled() {
+		return 0, origCol, nil
+	}
+	for _, col := range t.Columns {
+		if col.State != model.StatePublic {
+			nonPubCols = append(nonPubCols, col.ToInfo())
+			if origCol == nil && col.ChangeStateInfo != nil {
+				// find out the original column
+				for _, c := range t.Columns {
+					if c.Offset == col.ChangeStateInfo.DependencyColumnOffset {
+						origCol = c
+						break
+					}
+				}
+				if origCol == nil {
+					logutil.BgLogger().Warn("skip checksum since the original column is not found",
+						zap.Int64("tblID", t.meta.ID), zap.Any("cols", t.meta.Columns),
+						zap.Int("origColID", col.ChangeStateInfo.DependencyColumnOffset))
+					return 0, origCol, errors.New("cannot find the original column by change-state-info")
+				}
+			}
+		}
+	}
+	if n := len(nonPubCols); n > 1 {
+		logutil.BgLogger().Warn("skip checksum since the number of non-public columns is greater than 1",
+			zap.Int64("tblID", t.meta.ID), zap.Any("nonPubCols", nonPubCols))
+		return 0, origCol, nil
+	} else {
+		return 1 + n, origCol, nil
+	}
+}
+
+func (t *TableCommon) calcChecksums(data [][]rowcodec.ColData, buf []byte) ([]uint32, []byte) {
+	if len(data) == 0 {
+		return nil, buf
+	}
+	checksums := make([]uint32, len(data))
+	for i, cols := range data {
+		row := rowcodec.RowData{Cols: cols, Data: buf}
+		if !sort.IsSorted(row) {
+			sort.Sort(row)
+		}
+		checksum, err := row.Checksum()
+		buf = row.Data
+		if err != nil {
+			logutil.BgLogger().Warn("skip checksum due to encode error", zap.Int64("tblID", t.meta.ID), zap.Error(err))
+			return nil, buf
```

Comment:
Could the returned `buf` be already changed and corrupted?

Reply:
The returned `buf` is used to update the cached buf (for reduce memory allocation), its content doesn't matter, all functions that take a `buf` as arg will reset it by `buf = buf[:0]` (if it's not nil).

Reply:
How about returning `nil, nil` or adding some comments about it as there are no `mutable or immutable` references in golang, and we've encountered some cases in which the input memory was changed unexpectedly.

Reply:
The returned `buf` has the following properties, thus returning `nil, nil` somehow means invalid this buffer, or we can discard the nil? (do not update the persistent buf when it's nil)

- if cap of buf is enough, return the original buf, maybe with different len
- if cap of buf is not enough, return a new buf with enough capacity (by append(buf, ...)

The content of the returned `buf` is undefined actually, thus callers shouldn't rely on its content (at least for `calcChecksums`). `EncodeRow` also receives a bytes as row buf, and returns the row value contained in the original buf (or a new buf with enough capacity). It's callers' responsiblity to copy the returned bytes if they want to pesist it (membuf has a copy semantics.